src/c2ffi/asdf.lisp

```diff
@@ -0,0 +1,162 @@
+   (include-definitions :initarg :include-definitions
+                        :initform :all
+                        :accessor c2ffi-file/include-definitions)
+   (exclude-definitions :initarg :exclude-definitions
+                        :initform nil
+                        :accessor c2ffi-file/exclude-definitions))
+  (:default-initargs
+   :type nil)
+  (:documentation
+   "The input of this ASDF component is a C header file, plus configuration for the
+binding generation process. This header file will define the initial scope of
+the generation process, which can further be filtered by other configuration
+parameters.
+
+An external program based on clang/llvm called 'c2ffi' is then invoked with
+this header file to generate a json file for various platforms. The generation
+of the underlying platform's json file must succeed to continue, but the
+generation for the other platforms is allowed to fail
+\(see ENSURE-SPEC-FILE-EXISTS).
+
+It's advisable to deliver these json files with the project, so that users
+don't need to have c2ffi installed.
+
+Once/if the json file is available for the underlying platform, then the json
+file is used to generate a lisp file with CFFI definitions (see
+PROCESS-C2FFI-SPEC-FILE). This file will then be compiled as any other lisp
+file, except that it's will be stored in the fasl cache."))
+
+(defclass generate-op (downward-operation)
+  ())
+
+(defmethod input-files ((op generate-op) (c c2ffi-file))
+  (list (component-pathname c)))
+
+(defmethod input-files ((op compile-op) (c c2ffi-file))
+  (list (first (output-files 'generate-op c))))
+
+(defmethod component-depends-on ((op generate-op) (c c2ffi-file))
+  `((prepare-op ,c) ,@(call-next-method)))
+
+(defmethod component-depends-on ((op compile-op) (c c2ffi-file))
+  `((generate-op ,c) ,@(call-next-method)))
+
+(defmethod component-depends-on ((op load-source-op) (c c2ffi-file))
+  `((generate-op ,c) ,@(call-next-method)))
+
+(defmethod output-files ((op generate-op) (c c2ffi-file))
+  (let* ((input-file (first (input-files op c)))
+         (spec-file (spec-path input-file))
+         (generated-lisp-file (make-pathname :type "lisp"
+                                             :defaults spec-file)))
+    ;; this order here influences various other places
+    (values (list* (asdf:apply-output-translations generated-lisp-file)
+                   ;; If we unconditionally return the SPEC-FILE here as an output file,
+                   ;; then that makes ASDF to unconditionally redo the rest of the
+                   ;; generation and compilation.
+                   (unless (probe-file spec-file)
+                     (list spec-file)))
+            t)))
+
+(defmethod perform ((op generate-op) (c c2ffi-file))
+  (let* ((output-file (first (output-files op c)))
+         (input-file (first (input-files op c)))
+         (spec-file (let ((*c2ffi-executable* (if (slot-boundp c 'c2ffi-executable)
```

Comment:
This binding pattern for `*c2ffi-executable*` and `*trace-c2ffi` is here so that they can be dynamically bound around the loading of a cffi/c2ffi-using library, is it?